src/persist/src/indexed/mod.rs

```diff
@@ -681,6 +682,66 @@ impl<L: Log, B: BlobRead> Indexed<L, B> {
         }
         true
     }
+
+    /// WIP
+    pub fn meta(&mut self, meta_res: Result<ProtoMeta, Error>) {
+        debug_assert_eq!(self.validate_pending_empty(), Ok(()));
+
+        let new_meta = match meta_res {
+            Err(_) => {
+                // WIP inc a metric
+                return;
+            }
+            Ok(new_meta) => new_meta,
+        };
+        let new_state = AppliedState::new(new_meta.into());
+        let diffs = diff_metas(&self.state, &new_state, &self.blob);
+
+        for (id, event) in diffs {
+            if let Some(listeners) = self.listeners.get(&id) {
+                if listeners.len() == 1 {
+                    if let Err(crossbeam_channel::SendError(_)) = listeners[0].send(event) {}
+                } else {
+                    for listener in listeners.iter() {
+                        if let Err(crossbeam_channel::SendError(_)) = listener.send(event.clone()) {
+                        }
+                    }
+                }
+            }
+        }
+    }
+}
+
+fn diff_metas<B: BlobRead>(
```

Comment:
what do you do if the batches in new_meta don't exist anymore because they've been compacted away or something?

It kind of feels like if you fall back far enough behind you won't even be able to read back what you had previously sent from the old meta because all of the old batches are now gone, and in general its unclear to me what good options there are here without either:

- enforcing that no batches are physically removed from blob for some large time interval (say 24 hours) and then erroring if you observe a meta pointing to a batch that doesn't exist
- keeping a copy of everything sent locally somewhere, so you can always subtract that out from any new batches if needed
- something else??

Reply:
yeah i don't have a good story here yet. i've thought of a few things along the lines of "keep batches around for a while". another idea is that for the "production/production" case, we could allow read-only instances to hold back compaction. that doesn't work for the "production/staging" case though